libr/core/cmd_print.c

```diff
@@ -1613,40 +1625,40 @@ static void cmd_print_format(RCore *core, const char *_input, const ut8* block,
 				r_print_format (core->print, core->offset,
 					core->block, core->blocksize, name, mode, NULL, NULL);
 			}
+	err_name:
 			free (name);
 		}
 	} else {
-		/* This make sure the structure will be printed entirely */
+		/* Make sure the structure will be printed entirely */
 		const char *fmt = r_str_trim_head_ro (input + 1);
 		int struct_sz = r_print_format_struct_size (core->print, fmt, mode, 0);
 		int size = R_MAX (core->blocksize, struct_sz);
 		ut8 *buf = calloc (1, size);
-		if (!buf) {
-			eprintf ("cannot allocate %d byte(s)\n", size);
-			goto stage_left;
-		}
+		if (!buf) goto err_buf;
 		memcpy (buf, core->block, core->blocksize);
-		/* check if fmt is '\d+ \d+<...>', common mistake due to usage string*/
-		bool syntax_ok = true;
+
 		char *args = strdup (fmt);
-		if (!args) {
-			r_cons_printf ("Error: Mem Allocation.");
-			free (args);
-			goto stage_left;
-		}
+		if (!args) goto err_args;
+
+		/* check if fmt is '\d+ \d+<...>', common mistake due to usage string*/
 		const char *arg1 = strtok (args, " ");
 		if (arg1 && r_str_isnumber (arg1)) {
-			syntax_ok = false;
-			r_cons_printf ("Usage: pf [0|cnt][format-string]\n");
+			r_core_cmd_help (core, (const char *[]){
+				"Usage:", "pf [0|cnt][format-string]", "",
+				NULL
+			});
+			goto err_arg1;
 		}
+
+		r_print_format (core->print, core->offset,
+			buf, size, fmt, mode, NULL, NULL);
+	err_arg1:
 		free (args);
-		if (syntax_ok) {
-			r_print_format (core->print, core->offset,
-				buf, size, fmt, mode, NULL, NULL);
-		}
+	err_args:
 		free (buf);
+	err_buf:
```

Comment:
Let me put it this way. We need some way to get to the end of that if-then scope, because there's a cleanup routine at the end. Logically it would be the same as putting the label one level higher, right before the `free(input)` on the line below, but semantically it would be quite different. What's worse, it someone later would add some code in between, we'll be in trouble.

Better refactor that block into a function, so we can freely use `return` statement instead.

Reply:
yes when those situations happen it is better to find a better strategy like refactoring into separate functions. The pf code is quite complex and have been refactored a couple of times. But as many code in r2land, everything is subject to be improved. Im aware of how huge this refactoring can go so it's better to merge this when it's ready and continue further refactorings in a separate pr